crates/wit/src/commands/pull.rs

```diff
@@ -0,0 +1,324 @@
+            // Warn on unparsable root package; might unexpectedly be missing deps
+            if let Err(err) = UnresolvedPackage::parse_dir(&self.wit_dir) {
+                terminal.warn(format!("Couldn't parse root package: {err}"))?;
+            }
+            // No packages specified; pull missing dependencies
+            pkgs_state
+                .missing_deps()
+                .map(|pkg| {
+                    let name = format!("{}:{}", pkg.namespace, pkg.name).parse().unwrap();
+                    let version = pkg
+                        .version
+                        .as_ref()
+                        .map(|ver| ver.to_string().parse().unwrap());
+                    VersionedPackageName { name, version }
+                })
+                .collect::<Vec<_>>()
+        } else {
+            // Remove existing packages from given list
+            self.packages
+                .iter()
+                .filter(|pkg| !pkgs_state.satisfies(pkg))
+                .cloned()
+                .collect()
+        };
+
+        if packages.is_empty() {
+            terminal.status("Finished", "no missing packages; nothing to do")?;
+            return Ok(());
+        }
+        log::debug!("Packages to pull: {packages:?}");
+
+        let mut client = {
+            let mut config = ClientConfig::default();
+            config.namespace_registry("wasi", "bytecodealliance.org");
+            if let Some(file_config) = ClientConfig::from_default_file()? {
+                config.merge_config(file_config);
+            }
+            if let Some(registry) = self.registry.clone() {
+                config.default_registry(registry);
+            }
+            config.to_client()
+        };
+
+        for pkg in packages {
+            if pkgs_state.satisfies(&pkg) {
+                log::info!("Skipping {pkg}; resolved by previous pull?");
+                continue;
+            }
+            terminal.status("Resolving", format!("package {pkg}"))?;
+
+            match self.pull(&mut client, &pkg).await? {
+                Some((release, decoded)) => {
+                    let root_pkg = &decoded.resolve().packages[decoded.package()];
+
+                    let name = &pkg.name;
+                    let release_pkg = PackageName {
+                        namespace: name.namespace().to_string(),
+                        name: name.name().to_string(),
+                        version: Some(release.version.clone()),
+                    };
+                    terminal.status("Downloaded", format!("release {release_pkg}"))?;
+
+                    let version = root_pkg.name.version.clone();
+                    if let Some(wit_version) = &version {
+                        let release_version = &release.version;
+                        if wit_version != release_version {
+                            terminal.warn(format!("Release version {release_version} doesn't match WIT package version {wit_version}"))?;
```

Comment:
I'm curious, when would this happen? What is the user supposed to do with this information?

Reply:
Currently nothing enforces that the outer registry version (Warg release version or OCI tag) matches the inner WIT package version. This is a problem here because we only have the WIT version for the "already installed" set and can only use the registry version to resolve packages to pull.

I'm not sure how to resolve this yet, so the warning serves primarily to alert publishers that something is wrong.